File: app/src/main/java/org/wikiedufoundation/wikiedudashboard/course_list/view/CourseListFragment.java
package org.wikiedufoundation.wikiedudashboard.course_list.view;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import org.wikiedufoundation.wikiedudashboard.R;
import org.wikiedufoundation.wikiedudashboard.adapters.CourseListRecyclerAdapter;
import org.wikiedufoundation.wikiedudashboard.course_detail.common.view.CourseDetailActivity;
import org.wikiedufoundation.wikiedudashboard.course_list.data.ExploreCoursesResponse;
import org.wikiedufoundation.wikiedudashboard.course_list.presenter.CourseListPresenterImpl;
import org.wikiedufoundation.wikiedudashboard.course_list.provider.RetrofitCourseListProvider;
import org.wikiedufoundation.wikiedudashboard.helper.SharedPrefs;
import org.wikiedufoundation.wikiedudashboard.helper.ViewUtils;

import butterknife.BindView;
import butterknife.ButterKnife;


/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * Use the {@link CourseListFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class CourseListFragment extends Fragment implements CourseListView {
    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";

    // TODO: Rename and change types of parameters
    private String mParam1;
    private String mParam2;
    @BindView(R.id.tv_no_courses)
    TextView tv_no_courses;

    @BindView(R.id.progressBar)
    ProgressBar progressBar;

    @BindView(R.id.rv_order_list)
    RecyclerView recyclerView;

    private SharedPrefs sharedPrefs;
    private Context context;
    private CourseListPresenterImpl courseListPresenter;
    private CourseListRecyclerAdapter courseListRecyclerAdapter;
    public CourseListFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @param param2 Parameter 2.
     * @return A new instance of fragment ExploreFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static CourseListFragment newInstance(String param1, String param2) {
        CourseListFragment fragment = new CourseListFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
            mParam2 = getArguments().getString(ARG_PARAM2);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_explore_course_list, container, false);
        ButterKnife.bind(this, view);
        context = getContext();
        sharedPrefs = new SharedPrefs(context);
        tv_no_courses.setText(sharedPrefs.getCookies());
        courseListPresenter = new CourseListPresenterImpl(this, new RetrofitCourseListProvider());
        courseListRecyclerAdapter = new CourseListRecyclerAdapter(context, this);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(context);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(courseListRecyclerAdapter);

        courseListPresenter.requestDashboard(sharedPrefs.getCookies());
        return view;
    }

    @Override
    public void setData(ExploreCoursesResponse data) {
        Log.d("DashboardFragment: ", data.toString());
        if (data.getCourses().size() > 0) {
            recyclerView.setVisibility(View.VISIBLE);
            courseListRecyclerAdapter.setData(data.getCourses());
            courseListRecyclerAdapter.notifyDataSetChanged();
            tv_no_courses.setVisibility(View.GONE);
        }
        else {
            recyclerView.setVisibility(View.GONE);
            tv_no_courses.setVisibility(View.VISIBLE);
        }

    }

    @Override
    public void showProgressBar(boolean show) {
        if (show){
            progressBar.setVisibility(View.VISIBLE);
        }else {
            progressBar.setVisibility(View.GONE);
        }
    }

    @Override
    public void showMessage(String message) {
        ViewUtils.INSTANCE.showToast(context, message);
    }

    public void openCourseDetail(String slug) {
        Intent i = new Intent(context, CourseDetailActivity.class);
        i.putExtra("url", slug);
        startActivity(i);
    }
}

File: app/src/main/java/org/wikiedufoundation/wikiedudashboard/course_list/view/CourseListFragment.kt
package org.wikiedufoundation.wikiedudashboard.course_list.view

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView

import org.wikiedufoundation.wikiedudashboard.R
import org.wikiedufoundation.wikiedudashboard.adapters.CourseListRecyclerAdapter
import org.wikiedufoundation.wikiedudashboard.course_detail.common.view.CourseDetailActivity
import org.wikiedufoundation.wikiedudashboard.course_list.data.ExploreCoursesResponse
import org.wikiedufoundation.wikiedudashboard.course_list.presenter.CourseListPresenterImpl
import org.wikiedufoundation.wikiedudashboard.course_list.provider.RetrofitCourseListProvider
import org.wikiedufoundation.wikiedudashboard.helper.SharedPrefs
import org.wikiedufoundation.wikiedudashboard.helper.ViewUtils

import butterknife.BindView
import butterknife.ButterKnife


/**
 * A simple [Fragment] subclass.
 * Activities that contain this fragment must implement the
 * Use the [CourseListFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class CourseListFragment : Fragment(), CourseListView {

    private var mParam1: String? = null
    private var mParam2: String? = null
    private var tv_no_courses: TextView? = null
    private var progressBar: ProgressBar? = null
    private var recyclerView: RecyclerView? = null

    private var courseListPresenter: CourseListPresenterImpl? = null
    private var courseListRecyclerAdapter: CourseListRecyclerAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            mParam1 = arguments!!.getString(ARG_PARAM1)
            mParam2 = arguments!!.getString(ARG_PARAM2)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_explore_course_list, container, false)
        progressBar = view.findViewById(R.id.progressBar)
        tv_no_courses = view.findViewById(R.id.tv_no_courses)
        recyclerView = view.findViewById(R.id.rv_course_list)

        val context: Context? = getContext()
        val sharedPrefs : SharedPrefs? = SharedPrefs(context)
        tv_no_courses!!.text = sharedPrefs!!.cookies
        courseListPresenter = CourseListPresenterImpl(this, RetrofitCourseListProvider())
        courseListRecyclerAdapter = CourseListRecyclerAdapter(context!!, this)
        val linearLayoutManager = LinearLayoutManager(context)
        recyclerView!!.layoutManager = linearLayoutManager
        recyclerView!!.setHasFixedSize(true)
        recyclerView!!.adapter = courseListRecyclerAdapter

        courseListPresenter!!.requestDashboard(sharedPrefs.cookies!!)
        return view
    }

    override fun setData(data: ExploreCoursesResponse) {
        Log.d("DashboardFragment: ", data.toString())
        if (data.courses.size > 0) {
            recyclerView!!.visibility = View.VISIBLE
            courseListRecyclerAdapter!!.setData(data.courses)
            courseListRecyclerAdapter!!.notifyDataSetChanged()
            tv_no_courses!!.visibility = View.GONE
        } else {
            recyclerView!!.visibility = View.GONE
            tv_no_courses!!.visibility = View.VISIBLE
        }

    }

    override fun showProgressBar(show: Boolean) {
        if (show) {
            progressBar!!.visibility = View.VISIBLE
        } else {
            progressBar!!.visibility = View.GONE
        }
    }

    override fun showMessage(message: String) {
        ViewUtils.showToast(context!!, message)
    }

    fun openCourseDetail(slug: String) {
        val i = Intent(context, CourseDetailActivity::class.java)
        i.putExtra("url", slug)
        startActivity(i)
    }

    companion object {
        // TODO: Rename parameter arguments, choose names that match
        // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
        private val ARG_PARAM1 = "param1"
        private val ARG_PARAM2 = "param2"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param param1 Parameter 1.
         * @param param2 Parameter 2.
         * @return A new instance of fragment ExploreFragment.
         */
        // TODO: Rename and change types and number of parameters
        fun newInstance(param1: String, param2: String): CourseListFragment {
            val fragment = CourseListFragment()
            val args = Bundle()
            args.putString(ARG_PARAM1, param1)
            args.putString(ARG_PARAM2, param2)
            fragment.arguments = args
            return fragment
        }
    }
}// Required empty public constructor
